### PR TITLE
feat: replicate_append

### PR DESCRIPTION
**Pull request by BoltonBailey**
We currently have [List.replicate_add](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/List/Basic.html#List.replicate_add). I think it would be better to introduce `replicate_append` instead as the `symm` of this, and let it be a `simp` lemma. At the very least, we should have this be the main lemma, as grepping for `replicate_add` indeed shows that it is usually used in `<-` form.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Basic.lean [hunk 1]
```diff
@@ -437,12 +437,11 @@ theorem eq_replicate_length {a : α} : ∀ {l : List α}, l = replicate l.length
 
 #align list.eq_replicate List.eq_replicate
 
-theorem replicate_add (m n) (a : α) : replicate (m + n) a = replicate m a ++ replicate n a := by
```
Comment (eric-wieser) [inline Mathlib/Data/List/Basic.lean:440]: I don't really understand the motivation for this flip; morally this lemma is the same as `pow_add : a^(m+n) = a^m*a^n`, which I don't think anyone has suggested is backwards.

**Review by eric-wieser — CHANGES_REQUESTED** [review]

File: Mathlib/Data/List/Basic.lean (hunk 1 above)
Comment (BoltonBailey) [reply Mathlib/Data/List/Basic.lean:440]: Is pow add used in symm form more often than in forward form? If so, then I would absolutely suggest that the reverse be put into mathlib.

Comment (eric-wieser) [reply Mathlib/Data/List/Basic.lean:440]: I don't think looking at which direction a lemma is used in is a particularly good indication of which way around it should be. Most forward rewrites on a goal can be replaced with a backwards rewrite on the hypothesis, and we're certainly not trying to optimize purely for the minimum number of left arrows.

Comment (BoltonBailey) [reply Mathlib/Data/List/Basic.lean:440]: Among the benefits of doing orientation it this way in my mind are:
*If `rw`able lemmas are oriented in the direction they are most often used, then people will be able to guess their names faster by typing in what they expect the lemma name to be, based on the terms they expect to see in the left hand side.
* Proofs become more readable if one can place one's cursor before a rewrite and see from the lemma name what part of the proof is about to be rewritten.
* There is be some ineffable force that makes lemmas be used in a certain direction more frequently than in the other, and even if I'm not sure what it is, I would like mathlib to track it more explicitly.

Comment (BoltonBailey) [reply Mathlib/Data/List/Basic.lean:440]: I suppose the best counterargument in my mind is that the distributivity type lemmas all seem to be oriented the same way. If someone were familiar enough with the library, they might have noticed this and that might have influenced their expectation WRT the orientation. I'm not sure how that balances against the opportunity to capture this ineffable force.

Comment (eric-wieser) [reply Mathlib/Data/List/Basic.lean:440]: I think part of the reason that distributivity lemmas are the way that they are is that `foo a (bar b c) = bar (foo a b) (foo a c)` has a shorter name as `foo_bar` (forward) than as `foo_bar_foo` (backwards); combined with the fact that `mul_add` is in many ways the "obvious" way around, and the less obvious cases followed suit.

**Pull request closed without merging**